x-pack/plugins/data_enhanced/server/search/session/session_service.ts

```diff
@@ -25,21 +33,154 @@ import {
+}
+
 export class BackgroundSessionService implements ISessionService {
   /**
    * Map of sessionId to { [requestHash]: searchId }
    * @private
    */
-  private sessionSearchMap = new Map<string, Map<string, string>>();
+  private sessionSearchMap = new Map<string, SessionInfo>();
+  private internalSavedObjectsClient!: SavedObjectsClientContract;
+  private monitorTimer!: NodeJS.Timeout;
+
+  constructor(private readonly logger: Logger) {}
+
+  public async start(core: CoreStart, config$: Observable<ConfigSchema>) {
+    return this.setupMonitoring(core, config$);
+  }
+
+  public stop() {
+    this.sessionSearchMap.clear();
+    clearTimeout(this.monitorTimer);
+  }
+
+  private setupMonitoring = async (core: CoreStart, config$: Observable<ConfigSchema>) => {
+    const config = await config$.pipe(first()).toPromise();
+    if (config.search.sendToBackground.enabled) {
+      this.logger.debug(`setupMonitoring | Enabling monitoring`);
+      const internalRepo = core.savedObjects.createInternalRepository([BACKGROUND_SESSION_TYPE]);
+      this.internalSavedObjectsClient = new SavedObjectsClient(internalRepo);
+      this.monitorMappedIds();
+    }
+  };
+
+  /**
+   * Gets all {@link SessionSavedObjectAttributes | Background Searches} that
+   * currently being tracked by the service.
+   *
+   * @remarks
+   * Uses `internalSavedObjectsClient` as this is called asynchronously, not within the
+   * context of a user's session.
+   */
+  private async getAllMappedSavedObjects() {
+    const activeMappingIds = Array.from(this.sessionSearchMap.keys())
+      .map((sessionId) => `"${sessionId}"`)
+      .join(' | ');
+    const res = await this.internalSavedObjectsClient.find<BackgroundSessionSavedObjectAttributes>({
+      perPage: INMEM_MAX_SESSIONS, // If there are more sessions in memory, they will be synced when some items are cleared out.
+      type: BACKGROUND_SESSION_TYPE,
+      search: activeMappingIds,
+      searchFields: ['sessionId'],
+      namespaces: ['*'],
+    });
+    this.logger.debug(`getAllMappedSavedObjects | Got ${res.saved_objects.length} items`);
+    return res.saved_objects;
+  }
 
-  constructor() {}
+  private clearSessions = () => {
+    const curTime = moment();
+
+    this.sessionSearchMap.forEach((sessionInfo, sessionId) => {
+      if (
+        moment.duration(curTime.diff(sessionInfo.insertTime)).asSeconds() >
+        INMEM_TRACKING_TIMEOUT_SEC
+      ) {
+        this.logger.debug(`clearSessions | Deleting expired session ${sessionId}`);
+        this.sessionSearchMap.delete(sessionId);
+      } else if (sessionInfo.retryCount >= MAX_UPDATE_RETRIES) {
+        this.logger.warn(`clearSessions | Deleting failed session ${sessionId}`);
+        this.sessionSearchMap.delete(sessionId);
+      }
+    });
+  };
+
+  private async monitorMappedIds() {
+    this.monitorTimer = setTimeout(async () => {
+      try {
+        this.clearSessions();
+
+        if (!this.sessionSearchMap.size) return;
```

Comment:
Here because of early return we will not restart a timer

Reply:
Already fixed :-D